src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs

```diff
@@ -176,6 +176,33 @@ public void Reserve(int count)
         }
 
         /// <summary>
+        /// Reserves a region if the items' count can be predetermined; otherwise, adds the items to this builder.
+        /// </summary>
+        /// <param name="items">The items to reserve or add.</param>
+        /// <returns><c>true</c> if the items were reserved; otherwise, <c>false</c>.</returns>
+        /// <remarks>
+        /// If the items' count is predetermined to be 0, no reservation is made and the return value is <c>false</c>.
+        /// The effect is the same as if the items were added, since adding an empty collection does nothing.
+        /// </remarks>
+        public bool ReserveOrAdd(IEnumerable<T> items)
```

Comment:
@stephentoub Nvm, this type actually depends on `EnumerableHelpers.Linq.cs` which makes use of internal Linq interfaces. So there's no way to write unit tests for this type currently, unless maybe with `InternalsVisibleTo`. But either way that should probably be done in a separate PR bc some existing methods are outside the scope of this PR.